iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp

```diff
@@ -213,23 +217,73 @@ static LogicalResult buildModuleDescriptors(IREE::VM::ModuleOp &moduleOp,
   return success();
 }
 
-LogicalResult translateModuleToC(mlir::ModuleOp outerModuleOp,
-                                 llvm::raw_ostream &output) {
-  auto moduleOps = outerModuleOp.getOps<IREE::VM::ModuleOp>();
-  if (moduleOps.empty()) {
-    return outerModuleOp.emitError()
-           << "outer module does not contain a vm.module op";
-  }
-  auto moduleOp = *moduleOps.begin();
+// Adapted from BytecodeModuleTarget and extended by C specific passes
+static LogicalResult canonicalizeModule(IREE::VM::ModuleOp moduleOp) {
+  bool optimize = true;
+  bool stripDebugOps = true;
+
+  OwningRewritePatternList patterns;
+  ConversionTarget target(*moduleOp.getContext());
+  target.addLegalDialect<IREE::VM::VMDialect>();
+  target.addLegalOp<IREE::DoNotOptimizeOp>();
 
+  // Add all VM canonicalization patterns and mark pseudo-ops illegal.
   auto *context = moduleOp.getContext();
-  PassManager passManager(context);
+  for (auto *op : context->getRegisteredOperations()) {
+    // Non-serializable ops must be removed prior to serialization.
+    if (op->hasTrait<OpTrait::IREE::VM::PseudoOp>()) {
+      op->getCanonicalizationPatterns(patterns, context);
+      target.setOpAction(OperationName(op->name, context),
+                         ConversionTarget::LegalizationAction::Illegal);
+    }
+
+    // Debug ops must not be present when stripping.
+    // TODO(benvanik): add RemoveDisabledDebugOp pattern.
+    if (op->hasTrait<OpTrait::IREE::VM::DebugOnly>() && stripDebugOps) {
+      target.setOpAction(OperationName(op->name, context),
+                         ConversionTarget::LegalizationAction::Illegal);
+    }
+  }
+
+  if (failed(applyFullConversion(moduleOp, target, std::move(patterns)))) {
+    return moduleOp.emitError() << "unable to fully apply conversion to module";
+  }
 
+  PassManager passManager(context);
+  mlir::applyPassManagerCLOptions(passManager);
```

Comment:
For nested passes, I think you want runPipeline here, using the newish pipeline-in-a-pass infra. We have an example intree at https://github.com/google/iree/blob/main/iree/compiler/Dialect/HAL/Transforms/TranslateExecutables.cpp

Reply:
This then also affects the bytecode module target. So I'd suggest to change it in a follow up PR.